quantized conv: add support for graph mode BC testing, and increase coverage

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43651 ConvPackedParams: remove legacy format
* #43086 Adding a version serialization type to ConvPackedParam
* **#43524 quantized conv: add support for graph mode BC testing, and increase coverage**

Summary:

1. adds support for testing BC on data format and numerics for graph mode
quantized modules
2. using the above, adds coverage for quantized conv2d on graph mode

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23335222](https://our.internmc.facebook.com/intern/diff/D23335222)

vkuzo committed:
quantized conv: add additional bc test

Summary:

Adding coverage for quantized conv without bias in the backwards
compatibility tests

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
improve BC tests on "quantized conv: add additional bc test"


Summary:

Adding coverage for quantized conv without bias in the backwards
compatibility tests

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -69,6 +71,69 @@ def remove_prefix(text, prefix):
         self.assertEqual(qmodule_scripted(input_tensor), expected, atol=prec)
         self.assertEqual(qmodule_traced(input_tensor), expected, atol=prec)
 
+    def _test_op_graph(self, qmodule, subname=None, input_size=None, input_quantized=True,
```
looks like we are copying some code here, is it possible to dedup?

vkuzo replied:
sg, reused a bit more code, hope that works!

vkuzo commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -69,6 +71,69 @@ def remove_prefix(text, prefix):
+        # NB: we take __file__ from the module that defined the test
+        # class, so we place the expect directory where the test script
+        # lives, NOT where test/common_utils.py lives.
+        module_id = self.__class__.__module__
+        munged_id = remove_prefix(self.id(), module_id + ".")
+        test_file = os.path.realpath(sys.modules[module_id].__file__)
+        base_name = os.path.join(os.path.dirname(test_file),
+                                 "serialized",
+                                 munged_id)
+
+        subname_output = ""
+        if subname:
+            base_name += "_" + subname
+            subname_output = " ({})".format(subname)
+
+        input_file = base_name + ".input.pt"
+        state_dict_file = base_name + ".state_dict.pt"
+        scripted_module_file = base_name + ".scripted.pt"
+        traced_module_file = base_name + ".traced.pt"
+        expected_file = base_name + ".expected.pt"
```
yeah, there is potential to reuse some part of this code, it didn't seem worth it at first glance as it would increase complexity and the benefit would be small as not all of this can be reused.  I can take another look if you prefer.

jerryzh168 replied:
I don't have a strong preference, I just feel in general we should not be copying code.

vkuzo commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -69,6 +71,69 @@ def remove_prefix(text, prefix):
+        expected_file = base_name + ".expected.pt"
+
+        # only generate once.
+        if generate and qengine_is_fbgemm():
+            input_tensor = torch.rand(*input_size).float()
+            torch.save(input_tensor, input_file)
+
+            # convert to TorchScript
+            scripted = torch.jit.script(qmodule)
+            traced = torch.jit.trace(qmodule, input_tensor)
+
+            # quantize
+
+            def _eval_fn(model, data):
+                model(data)
+
+            qconfig_dict = {'': torch.quantization.default_qconfig}
+            scripted_q = torch.quantization.quantize_jit(
+                scripted, qconfig_dict, _eval_fn, [input_tensor])
+            traced_q = torch.quantization.quantize_jit(
+                traced, qconfig_dict, _eval_fn, [input_tensor])
+
+            torch.jit.save(scripted_q, scripted_module_file)
+            torch.jit.save(traced_q, traced_module_file)
+            torch.save(scripted_q(input_tensor), expected_file)
+
+        input_tensor = torch.load(input_file)
+        qmodule_scripted = torch.jit.load(scripted_module_file)
+        qmodule_traced = torch.jit.load(traced_module_file)
+        expected = torch.load(expected_file)
+        self.assertEqual(qmodule_scripted(input_tensor), expected, atol=prec)
+        self.assertEqual(qmodule_traced(input_tensor), expected, atol=prec)
```
for potentially reusing parts of this, it's different enough that I think the current way is cleaner given the readability benefits

vkuzo committed:
Update on "quantized conv: add support for graph mode BC testing, and increase coverage"


Summary:

1. adds support for testing BC on data format and numerics for graph mode
quantized modules
2. using the above, adds coverage for quantized conv2d on graph mode

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
rebase on "quantized conv: add support for graph mode BC testing, and increase coverage"


Summary:

1. adds support for testing BC on data format and numerics for graph mode
quantized modules
2. using the above, adds coverage for quantized conv2d on graph mode

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23335222](https://our.internmc.facebook.com/intern/diff/D23335222)

[ghstack-poisoned]

z-a-f reviewed (approved):
LGTM

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -69,6 +77,51 @@ def remove_prefix(text, prefix):
+            torch.save(scripted_q(input_tensor), expected_file)
+
+        input_tensor = torch.load(input_file)
+        qmodule_scripted = torch.jit.load(scripted_module_file)
```
Should we also have a test to catch regressions in the size of the quantized module files? We have had one issue before where the quantized model was storing both fp32 and quantized weights.

vkuzo replied:
I agree, that would be valuable as well.  It would probably be good for that live somewhere else since there could be things which don't break BC but change model size, and vice versa.  For this PR stack we are doing manual testing of file size, and it's a reduction.

raghuramank100 commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -69,6 +77,51 @@ def remove_prefix(text, prefix):
+        qmodule_scripted = torch.jit.load(scripted_module_file)
+        qmodule_traced = torch.jit.load(traced_module_file)
+        expected = torch.load(expected_file)
+        self.assertEqual(qmodule_scripted(input_tensor), expected, atol=prec)
```
Is there any value in comparing graph with eager expected outputs?

vkuzo replied:
There is definitely value in that, looks like that's already done in `test_quantize_jit.py`, so by those tests and the fact that both eager and graph are tested here for BC numerics, we can prove that eager vs graph condition is satisfied indirectly.

vkuzo committed:
Update on "quantized conv: add support for graph mode BC testing, and increase coverage"


Summary:

1. adds support for testing BC on data format and numerics for graph mode
quantized modules
2. using the above, adds coverage for quantized conv2d on graph mode

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23335222](https://our.internmc.facebook.com/intern/diff/D23335222)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized conv: add support for graph mode BC testing, and increase coverage"


Summary:

1. adds support for testing BC on data format and numerics for graph mode
quantized modules
2. using the above, adds coverage for quantized conv2d on graph mode

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23335222](https://our.internmc.facebook.com/intern/diff/D23335222)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized conv: add support for graph mode BC testing, and increase coverage"


Summary:

1. adds support for testing BC on data format and numerics for graph mode
quantized modules
2. using the above, adds coverage for quantized conv2d on graph mode

Test Plan:

```
python test/test_quantization.py TestSerialization.test_conv2d_nobias
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23335222](https://our.internmc.facebook.com/intern/diff/D23335222)

[ghstack-poisoned]
